### ISSUE TITLE
157d Who's Online no longer allows cart inspection

### ISSUE THREAD

**Issue by scottcwilson** [zencart/zencart#4710]
In 1.5.6, the session id was included in the link in who's online, allowing admins to inspect the customer's cart.

<img width="961" alt="156_whos_online" src="https://user-images.githubusercontent.com/4391638/161389809-763fe047-98f0-4919-9525-cbd6e624a209.png">

In 1.5.7, the session id no longer appears. 

<img width="991" alt="157" src="https://user-images.githubusercontent.com/4391638/161389835-5b41b11c-56e3-4f34-a82a-2ec59a65da7e.png">

**Comment by lat9** [comment]
Agreed; appears to be associated with this (https://github.com/zencart/zencart/commit/ae9706e6fb9aac6531c148fc0d5534601c69988c) commit.  Reverting to the zc157b version of the tool corrects the issue.

@mc12345678, could you verify, too?

I'll note that the associated change doesn't appear in the zc158 branch's history for the module.

**Comment by mc12345678** [comment]
> Agreed; appears to be associated with this ([ae9706e](https://github.com/zencart/zencart/commit/ae9706e6fb9aac6531c148fc0d5534601c69988c)) commit. Reverting to the zc157b version of the tool corrects the issue.
> 
> @mc12345678, could you verify, too?
> 
> I'll note that the associated change doesn't appear in the zc158 branch's history for the module.

What is the issue specifically? Is it that the path on screen doesn't have the session ID or is it that unable to identify what a customer has in the shopping cart? Or is there some third aspect not yet thrown into this issue? 

The issue referenced in the above commit ([ae9706e](https://github.com/zencart/zencart/commit/ae9706e6fb9aac6531c148fc0d5534601c69988c)) has to do with expired sessions not those that are still active and therefore as yet does not apply to this issue.

**Comment by lat9** [comment]
The issue is that, with the referenced change, admins can no longer click on an online-user to view their cart.  It doesn't matter what state the session is in; the customer-specific entry can't be clicked.

**Comment by mc12345678** [comment]
So, ZC 1.5.6c with PHP 7.3, I am not seeing the session id on the browser path. I really would not expect to see that parameter on the path unless there is a problem with sessions on the site and/or PHP 5.6 was in use (for some reason in sooo many cases it seemed to not support removing the session id). 

As far as seeing the contents of the cart, I disagree that it is not possible. The cart is displayed within the who's online tool and there is other information available to modify the display.  The base product and associated cost is already listed on the right side. 

Which customer's information is displayed can be updated. Clicking on the row that has the cart status, full name, ip address, entry time, last click time, and last url viewed will update the window to show the data for that customer.  Its not entirely difficult to add code to have all customer's carts shown on the right side instead of just the one-at-a-time, though that may also look confusing.

To accomplish what has been described with the current catalog side session code: either the code needs to attach the session parameter to the link, or the displayed session id is manually attached to the resulting link. My suggestion would be to "update" the cart display information to contain whatever it is the store owner wants to be able to see. As far as "taking over" the navigation for the cart customer, that's already a feature built in as an option to login as the customer.

As far as reverting to previous versions, the issue that was corrected outside of the minor modifications I added was that previous code caused the admin to get logged out while using the who's online tool. DrByte modified the whosOnline class to resolve that issue. Part of that resolution though too appears to have affected how the tool works, because: 
https://github.com/zencart/zencart/blob/40c161c8f5c0d58ac9e39d2dec479780d522e47e/admin/whos_online.php#L273

Includes a call to the function `zen_session_name`. It appears that code is intended to obtain the customer's session data, but In my testing within 1.5.7d, that function returned the session name associated with the admin. That seems to suggest the ability to report information related to an admin accessing a customer from the admin side of the store.  Otherwise that code snippet appears to no longer function as the method to append the customer's session data.

Looking for a summary? Making the link clickable to display the catalog side and able to access all information associated with that session is possible to be done, but appears like it shouldn't be offered/made easy. Display of information about the customer's cart, sure can be done, just needs to be incorporated with whatever formatting the particular store owner needs/wants.

**Comment by lat9** [comment]
To reproduce:
1. Sign into the zc157d admin.  You'll need an active site with a couple of sessions going, preferably with one that has an active cart.
2. Navigate to Tools :: Who's Online.
3. Click on one of the entries somewhere down the page, preferably one that has an active cart.
4. Notice that the right sidebox doesn't change; it still reflects the session information for the first item in the list.
5. Navigate away from Who's Online
6. Change /admin/whos_online.php to the zc157b version.
7. Repeat steps 2 and 3.
8. Notice that the right sidebox now reflects the cart items associated with the just-clicked session.

**Comment by mc12345678** [comment]
Had done so a few days ago without issue using just 1.5.7d as shown in the attached images. 

Original display: 
![Screenshot_20220403-104453_Chrome](https://user-images.githubusercontent.com/2072812/161830341-54a86bbc-8742-41ed-9201-1c5de793713f.jpg)

After clicking on the row that I previously described as the "first row" for the associated record: 
![20220403_145617](https://user-images.githubusercontent.com/2072812/161829302-3d80ca59-74bb-4048-93de-cca294b45e55.jpg)

Images were captured with two different browsers accessing the same site from the same device and each having one item on the cart with each item being different. Notice how the start of the session Identifier at the bottom matches the associated list item.

**Comment by proseLA** [comment]
> To reproduce:
> 
> 1. Sign into the zc157d admin.  You'll need an active site with a couple of sessions going, preferably with one that has an active cart.
> 2. Navigate to Tools :: Who's Online.
> 3. Click on one of the entries somewhere down the page, preferably one that has an active cart.
> 4. Notice that the right sidebox doesn't change; it still reflects the session information for the first item in the list.
> 5. Navigate away from Who's Online
> 6. Change /admin/whos_online.php to the zc157b version.
> 7. Repeat steps 2 and 3.
> 8. Notice that the right sidebox now reflects the cart items associated with the just-clicked session.

@lat9 i am not seeing this.  i am using the latest version of v157 (it is stamped v157c and is the v157d zip).  i have 2 sessions, and i am able to see both carts from that page whenever i click on a specific entry.  attached are 2 screenshots.  

am i missing something?

(and if you can get onto my dev server; by all means hijack my sessions....)

![image](https://user-images.githubusercontent.com/1095136/161866954-97ed7669-8a2e-4e89-bd5f-7982ad4a8d74.png)

![image](https://user-images.githubusercontent.com/1095136/161867054-6043d731-5870-44c3-919e-49d4d70935a9.png)

**Comment by lat9** [comment]
@proseLA, try clicking that other "Guest" online customer.  You'll see that it doesn't change from the prosela one.

**Comment by lat9** [comment]
OK, let me see what's different on the site I was testing.
